File: Sample/BasicSample/sample_cs/Browser/SampleBrowserLayer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

class SampleBrowserLayer : asd.Layer2D
{
	public static readonly int Columns = 3;
	public static readonly asd.Vector2DF ItemOffset = new asd.Vector2DF(632 / Columns, 150);

	private List<SampleItem> items;
	private asd.CameraObject2D camera;
	private SampleItem activeItem;
	private SampleBrowser browser;

	public asd.RectF CameraArea
	{
		get { return camera.Src.ToF(); }
	}
	public float TotalHeight { get; private set; }

	public SampleBrowserLayer(SampleBrowser browser, ISample[] samples)
	{
		Name = "BrowserLayer";
		this.browser = browser;
		items = new List<SampleItem>();

		var font = asd.Engine.Graphics.CreateDynamicFont("", 12, new asd.Color(255, 255, 255, 255), 1, new asd.Color(0, 0, 0, 255));
		int index = 0;
		foreach(var sample in samples)
		{
			var item = new SampleItem(sample, font);
			var x = ItemOffset.X * (index % Columns) + 8;
			var y = ItemOffset.Y * (index / Columns) + 8 + 20;
			item.Position = new asd.Vector2DF(x, y);
			AddObject(item);
			items.Add(item);

			index++;
		}

		var isThereJut = index % Columns == 0;

		if(isThereJut)
		{
			TotalHeight = ItemOffset.Y * (index / Columns) + 8 + 20;
		}
		else
		{
			TotalHeight = ItemOffset.Y * (index / Columns + 1) + 8 + 20;
		}


		camera = new asd.CameraObject2D();
		camera.Src = new asd.RectI(0, 0, 640, 480);
		camera.Dst = new asd.RectI(0, 0, 640, 480);
		AddObject(camera);
	}

	protected override void OnUpdated()
	{
		var rows = items.Count / Columns;

		if(items.Count % Columns != 0)
		{
			rows++;
		}

		var y = camera.Src.Y - asd.Engine.Mouse.MiddleButton.WheelRotation * 30;
		y = Math.Max(0.0f, y);
		y = Math.Min(rows * ItemOffset.Y - 480 + 24 + 60, y);
		camera.Src = new asd.RectI(
			camera.Src.X,
			(int)y,
			camera.Src.Width,
			camera.Src.Height);

		var mouse = new asd.CircleShape();
		mouse.Position = asd.Engine.Mouse.Position + new asd.Vector2DF(0, camera.Src.Y);
		mouse.OuterDiameter = 2;

		if(activeItem != null && !activeItem.Shape.GetIsCollidedWith(mouse))
		{
			activeItem.Disactivate();
			activeItem = null;
			browser.ShowInfo(null);
		}

		foreach(var item in items)
		{
			if(item.Shape.GetIsCollidedWith(mouse))
			{
				if(item != activeItem)
				{
					item.Activate();
					activeItem = item;
					browser.ShowInfo(item.Sample);
				}
				if(asd.Engine.Mouse.LeftButton.ButtonState == asd.ButtonState.Push)
				{
					browser.Selected = item.Sample;
				}
				break;
			}
		}
	}
}

File: Sample/BasicSample/sample_cpp/Browser/SampleBrowserLayer.cpp
#include "SampleBrowser.h"
#include "SampleBrowserLayer.h"
#include "SampleItem.h"

using namespace std;

SampleBrowserLayer::SampleBrowserLayer(SampleBrowser* browser, std::vector<SampleInfo>& samples)
	: m_items(std::vector<shared_ptr<SampleItem>>())
{
	SetName(asd::ToAString("BrowserLayer"));

	this->browser = browser;

	auto font = asd::Engine::GetGraphics()->CreateDynamicFont(
		asd::ToAString("").c_str(),
		12, asd::Color(255, 255, 255, 255),
		1, asd::Color(0, 0, 0, 255));
	int index = 0;
	for (auto& sample : samples)
	{
		auto item = make_shared<SampleItem>(sample, font);
		auto x = ItemOffsetX * (index % Columns) + 8;
		auto y = ItemOffsetY * (index / Columns) + 8 + 20;
		item->SetPosition(asd::Vector2DF(x, y));
		AddObject(item);
		m_items.push_back(item);
		++index;
	}

	auto haveJut = index % Columns == 0;
	if (haveJut)
	{
		m_totalHeight = ItemOffsetY * (index / Columns) + 8 + 20;
	}
	else
	{
		m_totalHeight = ItemOffsetY * (index / Columns + 1) + 8 + 20;
	}

	m_camera = make_shared<asd::CameraObject2D>();
	m_camera->SetSrc(asd::RectI(0, 0, 640, 480));
	m_camera->SetDst(asd::RectI(0, 0, 640, 480));
	AddObject(m_camera);
}

void SampleBrowserLayer::OnUpdated()
{
	auto rows = m_items.size() / Columns;

	if (m_items.size() % Columns != 0)
	{
		rows++;
	}

	auto y = m_camera->GetSrc().Y - asd::Engine::GetMouse()->GetMiddleButton()->GetRotation() * 30;
	y = max(0, y);
	y = min(rows * ItemOffsetY - 480 + 24 + 60, y);
	m_camera->SetSrc(asd::RectI(m_camera->GetSrc().X, (int)y, m_camera->GetSrc().Width, m_camera->GetSrc().Height));

	auto mouse = make_shared<asd::CircleShape>();
	mouse->SetPosition(asd::Engine::GetMouse()->GetPosition() + asd::Vector2DF(0, m_camera->GetSrc().Y));
	mouse->SetOuterDiameter(2);

	if (m_activeItem != nullptr && !m_activeItem->GetArea().GetIsCollidedWith(mouse))
	{
		m_activeItem->Disactivate();
		m_activeItem = nullptr;

		auto disabled = SampleInfo();
		disabled.isAvailable = false;
		browser->ShowInfo(disabled);
	}

	for (auto& item : m_items)
	{
		if (item->GetArea().GetIsCollidedWith(mouse))
		{
			if (item != m_activeItem)
			{
				item->Activate();
				m_activeItem = item;
				browser->ShowInfo(item->GetSample());
			}
			if (asd::Engine::GetMouse()->GetLeftButton()->GetButtonState() == asd::ButtonState::Push)
			{
				browser->Selected = item->GetSample();
			}
			break;
		}
	}
}

asd::RectF& SampleBrowserLayer::GetCameraArea() const
{
	return m_camera->GetSrc().ToF();
}

float SampleBrowserLayer::GetTotalHeight() const
{
	return m_totalHeight;
}